sprockets/mixins/amqp/amqp.py

```diff
@@ -0,0 +1,210 @@
+        if not self.connection or not self.channel:
+            LOGGER.error('Unable to connect to Rabbit')
+            raise pika.exceptions.AMQPError()
+        AMQPMixin.channel.basic_publish(exchange, routing_key, message,
+                                        properties)
+
+    @property
+    def _rabbit_params(self):
+        """Return a pika URLParameters object using the :envvar:`AMQP`
+        environment variable which identifies the Rabbit MQ server as a URL
+        ready for :class:`pika.connection.URLParameters`.
+
+        :rtype: pika.URLParameters
+
+        """
+        LOGGER.debug('URLParameters: %s',
+                     pika.URLParameters(os.environ['AMQP']))
+        return pika.URLParameters(os.environ['AMQP'])
+
+    def on_conn_close(self, reply_code, reply_text):
+        """Called when RabbitMQ has been connected to.
+
+        :param int reply_code: The code for the disconnect
+        :param str reply_text: The disconnect reason
+
+        """
+        LOGGER.warning('RabbitMQ has disconnected (%s): %s',
+                       reply_code, reply_text)
+        AMQPMixin.connection = None
+        self.amqp_ready.clear()
+        AMQPMixin.amqp_state = AMQP_STATES['disconnected']
+        self._connect_to_rabbitmq()
+
+    @gen.coroutine
+    def on_conn_open(self, connection):
+        """Called when RabbitMQ has been connected to.
+
+        :param pika.connection.Connection connection: The pika connection
+
+        """
+        LOGGER.info('RabbitMQ has created a connection')
+        AMQPMixin.connection = connection
+        AMQPMixin.connection.add_on_close_callback(self.on_conn_close)
+        yield self._create_new_channel()
+        self.amqp_ready.set()
+        LOGGER.info('RabbitMQ has connected')
+
+    def on_channel_open(self, channel):
+        """Called when the RabbitMQ accepts the channel open request.
+
+        :param pika.channel.Channel channel: The channel opened with RabbitMQ
+
+        """
+        LOGGER.info('Channel %i is opened for communication with RabbitMQ',
+                    channel.channel_number)
+        AMQPMixin.channel = channel
+        AMQPMixin.amqp_state = AMQP_STATES['connected']
+        self.amqp_channel_ready.set()
+
+    def on_channel_close(self, channel):
+        """Called when the RabbitMQ accepts the channel close request.
+
+        :param pika.channel.Channel channel: The channel closed with RabbitMQ
+
+        """
+        LOGGER.info('Channel %i is closed for communication with RabbitMQ',
+                    channel.channel_number)
+        AMQPMixin.channel = None
+        self.amqp_channel_ready.clear()
+        AMQPMixin.amqp_state = AMQP_STATES['disconnected']
+
+    def reset_amqp_state(self):
+        """ Clear the amqp_ready event, clear the amqp_channel_ready event and
+        set the AMQP_STATE to connecting.
+
+        """
+        AMQPMixin.amqp_state = AMQP_STATES['connecting']
```

Comment:
Is the default state "connecting"? Seems odd to have it connecting unless it is connecting